src/compiler/checker.ts

```diff
@@ -3605,10 +3604,18 @@ module ts {
             if (!type) {
                 type = unionTypes[id] = <UnionType>createObjectType(TypeFlags.Union | getWideningFlagsOfTypes(sortedTypes));
                 type.types = sortedTypes;
+                type.reducedType = noSubtypeReduction ? undefined : type;

```

Comment:
The `noSubtypeReduction` flag is there because it isn't possible to always do subtype reduction. The flag is true when creating a union type from a type node and when instantiating a union type. In both of those cases subtype reduction has to be deferred to properly support recursive union types.